### ISSUE TITLE
For CI - use Docker

### ISSUE THREAD

**Issue by weitjong** [urho3d/Urho3D#2348]
Continued discussion from https://github.com/urho3d/Urho3D/issues/743#issuecomment-404371445

**Comment by weitjong** [comment]
The CI build should now run like clockwork as it was long before. I hope this is the last time I need to touch/patch those all scripts. I think we have out grown Travis-CI. We should start to plan to move our CI logic to docker image(s) and run them in container. Once we achieve that, we are free to move to other greener pastures.

**Comment by weitjong** [comment]
Not having much time to work on this lately but at least after some trial and error, I have now working images locally. Will try to push them to Docker Hub after further testing and clean up. Hopefully sometime next week before I am preoccupied again with something else.

**Comment by weitjong** [comment]
Genting close now. I managed to setup the container to use a docker volume as ccache persistent storage between runs.

**Comment by weitjong** [comment]
Hit a bug in `docker-cli` which fortunately has already been fixed. The current stable version (18.06.x) didn't handle the `--env-file` option correctly. Need to install the latest test version (18.09.x) from the test repo.

**Comment by weitjong** [comment]
Native dockerized build environment is in. To test run it: `script/dockerized.sh native`. Windows batch file will be provided later, sorry.

**Comment by weitjong** [comment]
MinGW dockerized build environment is in the Docker Hub now.

**Comment by weitjong** [comment]
RPI dockerized build environment is ready.

**Comment by weitjong** [comment]
The ARM docker image contains two compiler toolchains (one 64-bit and one 32-bit), defaulted to 64-bit. Use the `URHO3D_64BIT` environment variable to choose between them during runtime, e.g. for 32-bit build

```
URHO3D_64BIT=0 script/dockerized.sh arm
```

**Comment by weitjong** [comment]
The last docker image for the Web build is also ready. Not tested it much yet though.

Next step is to take advantage of those docker images in our CI.

**Comment by weitjong** [comment]
It looks like docker-ce repo for Ubuntu Trusty (both in stable and edge) only contains 18.06.x, which does not handle `--env-file` option correctly (as I commented before). Sadly I have to stop my work here until things have caught up.

**Comment by weitjong** [comment]
I think I may have found a workaround to the root cause of the issue, i.e. shortlist the env-vars before passing them to Docker CLI.

**Comment by weitjong** [comment]
Created a new branch for the CI script migration to use the docker images. There are a few more things I need to prepare/fix in the existing CMake build scripts first before attempting the migration.

**Comment by ArnisLielturks** [comment]
@weitjong How about creating a docker Hub for Urho3D docker containers? Could be useful for others to get the build environment up and running in only few steps. I might also help with this.

**Comment by ArnisLielturks** [comment]
Oh, cool! Didn't noticed. Thanks for doing that!

**Comment by weitjong** [comment]
I think I should be able to wrap this one up by end of the week. The whole thing works better than I expected.

**Comment by weitjong** [comment]
Another thing I observe is that the docker hub "pull" statistics counts each pull from our own CI build. Bummer.

**Comment by weitjong** [comment]
I have decided to retain the CI build logic to split the jobs across different short-lived CI mirror branches. Although using the dockerized build environment that simplifies the ".travis.yml" a lot, it is now possible to group all the CI build jobs into just one huge build matrix. The benefit of retaining the logic are:
- we can have separate Travis-CI build statuses for each platform.
- PR is first checked by a small subset of CI jobs, not causing undue backlog to us and other Travis-CI users.
- When trouble-shooting, commits can be directly push to the CI mirror branch, again not causing undue backlog as each branch only has a subset of the CI jobs.

In future we would probably want to reorganize the primary CI to contain the most important build configurations for each platform together, and then we could have Linux native CI build and Linux MinGW Ci build separated into two more CI mirror branches. I will only do this after we have refactored the "https://urho3d.github.io/" main website. Currently the footer section does not have enough space there.

**Comment by weitjong** [comment]
The cache misses by 1 issue was caused by the generated source file for the LuaScript subsystem. It erroneously refers to dummy "Precompiled.h" in the build tree which has its timestamp touched each time.

**Comment by weitjong** [comment]
Finally narrow down the suspect for the cache miss by 1 issue. Still, it is related to the precompiled header. Turning off the URHO3D_PCH build option and the cache hits 100% on repeat build from scratch. Will investigate this further when I have time later.

**Comment by weitjong** [comment]
After the investigation I don't find there is anything wrong with how the PCH is currently being handled, i.e. the cache misses due to PCH doe snot need to be fixed. While checking the logs, I find out the reason why the Debug build config for the native CI build is unusually long and fixes that instead. And since it should be fast now, I added two more Debug build config CI for MinGW.

**Comment by weitjong** [comment]
Yesterday I changed the docker image for the Web dockerized build environment to activate the toolchain during docker image build time so that it does not thrash the emscripten-cache during runtime as it did before. Note that the container runtime is actually the urho3d project (or downstream project) build time. And with that, I added `.emscripten_cache` as one of the directory to be uploaded to S3 in `.travis.yml`. Hope it will pay off.

I have also just adjusted the CI build jobs for Android to clean up the *.lock and other marker files in order to avoid unnecessary gradle cache upload in case those are the only files being changed during the build. This should be my last commit for this branch. Finger cross.

I will merge this branch to master after all the CI finish with favorable result. Due to how our build scripts works, there could still be teething problem that will only show up in the master branch only, like documentation building and build artifact uploading.
